Python/Module6_Testing/Intro_to_Testing.md

```diff
@@ -0,0 +1,532 @@
+----> 1 assert 2 < 1
+
+AssertionError: 
+
+# we can include an error message with our assertion
+>>> assert 0 in [1, 2, 3], "0 is not in the list"
+---------------------------------------------------------------------------
+AssertionError                            Traceback (most recent call last)
+<ipython-input-8-e72fb36dc785> in <module>
+----> 1 assert 0 in [1, 2, 3], "0 is not in the list"
+
+AssertionError: 0 is not in the list
+```
+
+The general form of an assertion statement is:
+
+```python
+assert <expression> [, <error-message>] 
+```
+
+When an assertion statement is executed, the built-in `bool` function is called on the object that is returned by `<expression>`; if `bool(<expression>)` returns `False`, then an `AssertionError` is raised.
+If you included a string in the assertion statement - separated from `<expression>` by a comma - then this string will be printed as the error message.
+
+See that the assertion statement: 
+```python
+assert expression, error_message
+```
+
+is effectively shorthand for the following code (barring some additional details):
+
+```python
+# long-form equivalent of: `assert expression, error_message`
+if bool(expression) is False:
+    raise AssertionError(error_message)
+```
+
+<!-- #endregion -->
+<!-- #region -->
+<div class="alert alert-info"> 
+
+**Reading Comprehension: Assertions**
+
+Given the following objects:
+
+```python
+a_list = []
+a_number = 22
+a_string = "abcdef"
+```
+
+Write two assertion statements with the respective behaviors:
+
+- asserts that `a_list` is _not_ empty
+- asserts that the number of vowels in `a_string` is less than `a_number`; include and error message that prints the actual number of vowels
+
+</div>
+<!-- #endregion -->
+
+<!-- #region -->
+#### What is the Purpose of an Assertion?
+In our code, an assertion should be used as _a statement that is true unless there is a bug our code_.
+It is plain to see that the assertions in `test_count_vowels_basic` fit this description.
+However, it can also be useful to include assertions within our source code itself.
+For instance, we know that `count_vowels` should always return a non-negative integer for the vowel-count, and that it is illogical for this count to exceed the number of characters in the input string.
+We can explicitly assert that this is the case:
+
+```python
+# an example of including an assertion within our source code
+
+def count_vowels(x: str, include_y: bool = False) -> int:
+    vowels = set("aeiouAEIOU")
+    if include_y:
+        vowels.update("yY")
+    count = sum(1 for char in x if char in vowels)
+    
+    # This assertion should always be true: it is asserting that 
+    # the internal logic of our function is correct
+    assert isinstance(count, int) and 0 <= count <= len(x)
+    return count
+```
+
+Note that this assertion *is not meant to check if the user passed bad inputs for* `x` *and* `include_y`.
+Rather, it is meant to assert that our own internal logic holds true.
```

Comment:
Might be useful to mention the `-O` argument and `PYTHONOPTIMIZE=1` var - the *reason* assertions should never be used for validation logic is that they may or may not run.

That also makes it important to avoid depending on side-effects!